classes/actions.php

```diff
@@ -220,10 +234,14 @@ public static function duplicate_to_course(array $modules, int $targetcourseid,
             $sourcecoursemaxsectionnum = max(array_map(function($mod) use ($sourcemodinfo) {
                 return $sourcemodinfo->get_cm($mod->id)->sectionnum;
             }, $modules));
-            // We need to subtract 1 because there is section 0.
-            $targetcoursemaxsectionnum = count($targetmodinfo->get_section_info_all()) - 1;
 
-            for ($i = 0; $i < $sourcecoursemaxsectionnum - $targetcoursemaxsectionnum; $i++) {
+            $targetformatopt = $targetformat->get_format_options();
+            if (isset($targetformatopt['numsections'])) {
+                // Update course format setting to prevent orphaned sections.
+                update_course((object)array('id' => $targetcourseid, 'numsections' => $sourcecoursemaxsectionnum));
```

Comment:
Well yes, I didn't have in mind how we deal with the case of a user having `moodle/course:restoretargetimport`, but not `moodle/course:update`.

I missed out on that when I implemented the feature at first, but we could fix it right here:

I'd suggest to suppress the option "Keep sections" in the section select form if there are course modules with a higher source section number than we have sections in the target course and the user does not have the `moodle/course:update` cap.
Also, in this code line, we need to check for `moodle/course:update` and return, if we do not have this permission. So nothing would be restored.

Another idea would be, that we just add all course modules with higher source section num than available in the target course to the first or last section in the target course. But that feels a bit strange from UX side and makes things too complicated IMHO as we are discussing an edge case (well...not so edgy for our instance to be honest, but in general). Users without `moodle/course:update` can just import to a specific existing section and then do the sorting afterwards with `block_massaction`